### DISCUSSION TITLE
Mask/blur text in a document

### DISCUSSION THREAD

**Discussion by motdotla** [microsoft/vscode-discussions#104] [Extension Development QnA] [answered]
I would like to add secret masking to our Dotenv official extension. [1]

**Scenario**: For example, take a .env file like this opened in Code:

```
S3_KEY=secret_12345
STRIPE_KEY=prd_87643
```

If a user accidentally opens up that file while someone is watching (it might be on a screenshare or a screen recording) they've now put themselves and/or their organization at risk. They have to rotate those keys immediately to be safe.

**Solution**: I'd like to add a masking/blur feature, so that when they open their .env file anyone watching would not see the values. Instead, it would look something like this:

```
S3_KEY=************
STRIPE_KEY=************
```

When the user hovers over the masked/blurred value only then will the value be displayed.

**Question**: Is there a mechanism in VSCode extensions methods to accomplish this?

[1] https://marketplace.visualstudio.com/items?itemName=dotenv.dotenv-vscode&ssr=false#overview

**Comment by r3m0t** [comment, marked as answer]
You can set the color of the text to equal the backgroundColor using https://code.visualstudio.com/api/references/vscode-api#DecorationRenderOptions so it will appear as a solid block. You could also look at using `opacity` .

You can add an action above the first line of the .env to uncensor using https://code.visualstudio.com/api/references/vscode-api#CodeLensProvider%3CT%3E .

It might be possible purely with syntax highlighting or semantic tokens highlighting but I'm not familiar with that.

**Reply by motdotla** [reply]
Thank you, these are great suggestions. I'll start there and report back.

**Reply by motdotla** [reply]
Thanks again. This is starting to come together.

![mask](https://user-images.githubusercontent.com/3848/187752365-c0b918bd-af9d-4d0c-915e-cd9bae9b5856.gif)

**Reply by motdotla** [reply]
One minor issue. It takes the decorations a split second to render. Look at this video that I've slowed down to 10% time to show the split second rendering.

https://user-images.githubusercontent.com/3848/187802468-9bef735f-2823-4ad1-a4fc-d203d6fe703a.mp4


Is there a method to apply decorations BEFORE a file is opened? I didn't see one.

**Reply by r3m0t** [reply]
I didn't think of that. VS Code is designed to be responsive regardless of how inefficient any extension might be, therefore it is probably not possible to fix this with this API.

You could look at registering a custom editor instead of letting the normal text editor open. https://code.visualstudio.com/api/extension-guides/custom-editors

Or have a look at syntax highlighting as I mentioned before.

**Reply by motdotla** [reply]
Thank you. This turned out well. I combined both - to eliminate the flash and still maintain the blacked-out 'redacted' look.

I wish the VSCode team would implement backgroundColor formatting for semantic tokens in textMate themes. Code would be much simpler.
